AccountingSystem/Client/DataStoreClient.py

```diff
@@ -69,29 +72,53 @@ def __getRPCClient( self ):
       return RPCClient( "Accounting/DataStore", setup = self.__setup, timeout = 3600 )
     return RPCClient( "Accounting/DataStore", timeout = 3600 )
 
-  @gAccountingSynchro
   def commit( self ):
     """
     Send the registers in a bundle mode
     """
     rpcClient = self.__getRPCClient()
     sent = 0
-    while len( self.__registersList ) > 0:
-      registersToSend = self.__registersList[ :self.__maxRecordsInABundle ]
-      retVal = rpcClient.commitRegisters( registersToSend )
-      if retVal[ 'OK' ]:
-        self.__lastSuccessfulCommit = time.time()
-      else:
-        if self.__failoverEnabled and time.time() - self.__lastSuccessfulCommit > self.__maxTimeRetrying:
-          gLogger.verbose( "Sending accounting records to failover" )
-          result = _sendToFailover( retVal[ 'rpcStub' ] )
-          if not result[ 'OK' ]:
-            return result
+    
+    # create a local reference and prevent other running commits
+    # to take the same data second time 
+    self.__registersListLock.acquire()
+    registersList = self.__registersList
+    self.__registersList = []
+    self.__registersListLock.release()
+    
+    try:
+      while len( registersList ) > 0:
+        registersToSend = registersList[ :self.__maxRecordsInABundle ]
+        retVal = rpcClient.commitRegisters( registersToSend )
+        if retVal[ 'OK' ]:
+          self.__lastSuccessfulCommit = time.time()
         else:
-          return S_ERROR( "Cannot commit data to DataStore service" )
-      sent += len( registersToSend )
-      del( self.__registersList[ :self.__maxRecordsInABundle ] )
-    return S_OK( sent )
+          if self.__failoverEnabled and time.time() - self.__lastSuccessfulCommit > self.__maxTimeRetrying:
+            gLogger.verbose( "Sending accounting records to failover" )
+            result = _sendToFailover( retVal[ 'rpcStub' ] )
+            if not result[ 'OK' ]:
+              return result
+          else:
+            return S_ERROR( "Cannot commit data to DataStore service" )
+        sent += len( registersToSend )
+        del( registersList[ :self.__maxRecordsInABundle ] )
+    finally:
+      # if something is left because of an error return it to the main list
+      self.__registersList.extend(registersList)
```

Comment:
If you mean self.__registersList.extend(registersList) then, as far as I know, it doesn't need an additional lock. Please see here for an explanation:
http://effbot.org/pyfaq/what-kinds-of-global-value-mutation-are-thread-safe.htm